cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -530,7 +550,8 @@ Future<ZooKeeperReconciler> zooKeeperReconciler()   {
                                 versionChange,
                                 oldStorage,
                                 currentReplicas,
-                                clusterCa
+                                clusterCa,
+                                kafkaMetadataStateManager
```

Comment:
Why don't you pass just the state instead of the manager?

Reply:
The `ZooKeeperReconciler` class needs to call the `maybeDeleteZooKeeperControllerZnode` method on the manager instance during its reconcile (it happens within the reconciler `maybeDeleteControllerZnode` method).

Reply:
I think I commented on it later. I'm not sure why it should belong there and not into some `MegrationUtils` class or something.

Reply:
This would need the internal state of the FSM + if the annotation `disabled` is applied. I extracted a `KafkaMetadataStateManager.isRollingBack()` for it. If we want to move the deletion method to a MigrationUtils, we could need just a boolean to be passed here which represents if rolling back is going on or not.

Reply:
I have applied the refactoring described above.